### ISSUE TITLE
E231: Check misses trailing comma in function or class calls that do not end with whitespace

### ISSUE THREAD

**Issue by lfiedler** [PyCQA/pycodestyle#980]
Trailing commas in function calls, e.g. `print(1,2,)` that do not end with a white space are not flagged by pycodestyle.
From reading the corresponding [PEP8](https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas) I would expect that pycodestyle would flag this.

Here's a small example:
```python
def f(a, b):
    return a+b


class A:
    def __init__(self, a):
        self.a = a


if __name__ == "__main__":
    A(f(1, 2,),)
    print(f(1, 2,))
    print(1, 2,)
```
Although this is valid python code I'd expect pycodestyle to flag all of the three last lines, but it doesn't.
Is this behaviour expected?

Environment:
- Python 3.8.5
- conda 4.9.2
- pycodestyle 2.6.0

**Comment by asottile** [comment]
pycodestyle is an old tool and pep8 changes a lot so they're not always in sync (and even when they could be, sometimes pep8 has flip-flopped on an issue or is directly contradictory with itself so it's impossible to follow to a T).  this particular part of pep8 is [relatively new (2017)](https://github.com/python/peps/commit/c78ccb3dcb52c15b5177f19fa92fa97750e11b70#diff-db417657713fab81e79cb425b07ca55ecb11fc01fc52627e2756d450e8ac6f04) and so it's not surprising that pycodestyle has nothing for it.

additionally it's tricky to get correct, in your example `print(1, 2,)` it's hard to know whether this is a tuple or function call (since `pycodestyle` does not access the parse tree for performance reasons).

you might be interested in some tools which are built around handling these such as [add-trailing-comma](https://github.com/asottile/add-trailing-comma) or [black](https://github.com/psf/black) -- both of which are probably better suited to handle this than pycodestyle is

**Comment by lfiedler** [comment]
Thanks for the reply and also for the hints towards the plugin. I'll have a look.

Ironically, I was stumbling over this while using flake8, which, I am assuming, uses pycodestyle, in conjunction with black (which produces these kind of lines under certain circumstances, at least for the version I am using. But that does not belong here.).

**Comment by lfiedler** [comment]
I am not sure I understand your remark about `print(1, 2,)`: Even if this was a tuple, i.e. `(1, 2,)` this shouldn't be allowed.

**Comment by asottile** [comment]
consider `print(1,)`

**Comment by lfiedler** [comment]
I see, thanks.
